### PR TITLE
Generic interface

### PR DESCRIPTION
**Pull request by waydegg**
### Summary

This PR covers:
- updating all blocking code to be async
- creating a standard message format between the core library and client libraries
- moving all terminal-specific logic and dependencies under `mentat/terminal`

Things that will be covered in a later PR:
- A JSON RPC server with an LSP-compantible API that any editors with language server support can connect to 
- A `SessionManager` to coordinate multiple, concurrent chats

### To-do

- [ ] merge with main and resolve merge conflicts
- [ ] validate `cprint` colors received from core
- [ ] check/test for race conditions b/w core and terminal client (specifically w/ inputs and interrupts)
- [ ] fix test cases 
- [ ] create existing test cases
  - [ ] core library
  - [ ] terminal client
- [x] fix misc typing errors

### CODE REVIEW

**Review by granawkins — COMMENTED** [review]
Generally it looks great - seems to check all the boxes we identified with the concepts and what not. Nicely done 👏

**Comment by biobootloader** [comment]
I'm confused about why Github is showing a bunch of changes that have already been merged as part of the diff for this PR. Like my change here for PyPI going from `mentat-ai` to `mentat` is showing up in this PR:

image

**Review by PCSwingle — COMMENTED** [review]
Really love this! I'll take a closer look once it gets caught up to main, but looks good to me so far! Left a few questions on a couple things I wasn't sure about.

File: mentat/session_stream.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+from enum import Enum
+import logging
+from typing import Any, AsyncGenerator, Dict, List, cast
+from uuid import UUID, uuid4
+
+from .broadcast import Broadcast
+
+logger = logging.getLogger()
+
+_SESSION_STREAM: ContextVar[SessionStream] = ContextVar("mentat:session_stream")
```
[the remark this answers is not in the document]
Comment (waydegg) [reply mentat/session_stream.py:16]: Yeah we might want to do this with the entire `session` instance at some point

File: mentat/session_input.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    close the channel after receiving the input
+    """
+    stream = get_session_stream()
+
+    # data = {"type": "collect_user_input"}
+    # message = await stream.send(data)
+    # response = await stream.recv(f"default:{message.id}")
+
+    message = await stream.send("", channel="input_request")
+    response = await stream.recv(f"input_request:{message.id}")
```
[the remark this answers is not in the document]
Comment (waydegg) [reply mentat/session_input.py:23]: Totally. Adding more keys/values to the `StreamMessage` in the future won't break anything with how it's setup atm

**Comment by jakethekoenig** [comment]
I added `ipdb` to `requirements.txt`

File: mentat/terminal/client.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            await asyncio.sleep(0.1)
+
+    async def _run(self):
+        try:
+            self._init_signal_handlers()
+            await self._startup()
+            await self._main()
+            await self._shutdown()
+        # NOTE: if an exception is caught here, the main process will likely still run
+        # due to background ascynio Tasks that are still running
```
Comment (PCSwingle) [inline mentat/terminal/client.py:166]: nit: asyncio

**Comment by granawkins** [comment]
It looks like we lost the call to `setup_api_key()`, so I added it back in.

> 1. ctrl-c while the model is streaming closes mentat completely instead of returning to prompt.
> 2. Colors aren't working for me in streamed output removal. e.g. see the following conversation with coloring symbols instead of colors:
> 3. When mentat encounters an error e.g. the model doesn't correctly respond with line numbers then mentat starts streaming all the logs to stdout. Not sure if that's the intended behavior.

I'm setting the same errors as @jakethekoenig. I also got some logging to the terminal when running a diff:
```
granthawkins@Grants-MacBook-Air mentat % mentat --diff HEAD~1                                     
Diff annotations:
 ─•─ HEAD~1: a3063419: Fix logging config git conflict token | 2 files | 1 lines

DEBUG:root:Skipping /Users/granthawkins/mentat/.gitignore. Reason: lexer not found
Files included in context:
mentat
├── * .gitignore
└── mentat
    └── session.py
INFO:root:StreamMessage(id=UUID('71ba8857-3bb6-466f-9411-b3dce0a2d980'), channel='default', source=<StreamMessageSource.SERVER: 'server'>, data='\nIncluding CodeMap (filepaths only)', extra={'color': 'green'}, created_at=datetime.datetime(2023, 9, 28, 23, 31, 2, 618374))
INFO:root:StreamMessage(id=UUID('9651a3bf-e483-48da-8ca7-23c6a84470be'), channel='default', source=<StreamMessageSource.SERVER: 'server'>, data='File and prompt token count: 5979 / 8192', extra={'color': 'cyan'}, created_at=datetime.datetime(2023, 9, 28, 23, 31, 2, 621870))
INFO:root:StreamMessage(id=UUID('470117d2-e6c3-43e5-a705-672a525a62b0'), channel='default', source=<StreamMessageSource.SERVER: 'server'>, data="Type 'q' or use Ctrl-C to quit at any time.", extra={'color': 'cyan'}, created_at=datetime.datetime(2023, 9, 28, 23, 31, 2, 621913))
INFO:root:StreamMessage(id=UUID('d3598fd7-8925-4043-a805-c6fcc927b09d'), channel='default', source=<StreamMessageSource.SERVER: 'server'>, data='What can I do for you?', extra={'color': 'light_blue'}, created_at=datetime.datetime(2023, 9, 28, 23, 31, 2, 621932))
INFO:root:StreamMessage(id=UUID('a3f188bb-aabb-437b-96e2-7e58550f218f'), channel='input_request', source=<StreamMessageSource.SERVER: 'server'>, data='', extra={}, created_at=datetime.datetime(2023, 9, 28, 23, 31, 2, 621953))

Including CodeMap (filepaths only)
File and prompt token count: 5979 / 8192
Type 'q' or use Ctrl-C to quit at any time.
What can I do for you?
```

**Comment by waydegg** [comment]
Thanks for all of the comments everyone! I'm going to close and re-open this PR in an effort to get Github to show the proper git history (this might not be a github thing but we'll quickly find out lol).

**Pull request closed without merging**